### PR TITLE
clang-format for src/analysis

### PR DESCRIPTION
**Pull request by 3nids**
Here are the changes brought to allow using clang-format alongside with our sip code:
* includes in `TypeHeaderCode` is now achieved by using `SIP_TYPEHEADER_INCLUDE`
* `%MethodCode` and other directives are now commented (I think I would move them to using a macro `SIP_MethodCode` when the whole code base has already been moved to clang-format)
* `SIP_SKIP` should be placed before the definition of a method in the header (otherwise it's moved to next line by clang-format)

### CODE REVIEW

**Comment by 3nids** [comment]
This is ready for review and I would like to avoid waiting too long for a merge since rebasing is not funny :)

File: python/PyQt6/core/auto_generated/raster/qgsrasterinterface.sip.in [hunk 1]
```diff
@@ -109,24 +109,21 @@ Base class for processing filters like renderers, reprojector, resampler etc.
...
-// QgsRasterRenderer subclasses
-#include <qgshillshaderenderer.h>
-#include <qgsmultibandcolorrenderer.h>
-#include <qgspalettedrasterrenderer.h>
-#include <qgssinglebandcolordatarenderer.h>
-#include <qgssinglebandgrayrenderer.h>
-#include <qgssinglebandpseudocolorrenderer.h>
-#include <qgsrastersinglecolorrenderer.h>
-#include <qgsrastercontourrenderer.h>
+#include "qgsbrightnesscontrastfilter.h"
```
Comment (troopa81) [inline python/PyQt6/core/auto_generated/raster/qgsrasterinterface.sip.in:112]: According to your comment, are we not suppose to use SIP_TYPEHEADER_INCLUDE

Comment (3nids) [reply python/PyQt6/core/auto_generated/raster/qgsrasterinterface.sip.in:112]: yes, but in the header, it's translated in sip file

**Comment by troopa81** [comment]
> includes in TypeHeaderCode is now achieved by using SIP_TYPEHEADER_INCLUDE

Not the case for all files it seems, see [qgsprocessingparameter](https://github.com/qgis/QGIS/blob/6302af30a30eace54fc7c4673f6937b7527b2a1b/src/core/processing/qgsprocessingparameters.h#L338) for instance. Is it intentionnal ?

> %MethodCode and other directives are now commented (I think I would move them to using a macro SIP_MethodCode when the whole code base has already been moved to clang-format)

That mean that we would have a macro also for MappedType/ConvertFromTypeCode/End/ConvertToTypeCode see [qgsattributes](https://github.com/qgis/QGIS/blob/6302af30a30eace54fc7c4673f6937b7527b2a1b/src/core/qgsattributes.h#L146). ?

I'm not sure about commenting them either as it is IMHO a little bit error prone. Comment followed with % would become SIP instruction and other comment would stay a comment.

can we not disable clang-format around given lines like with clang-tidy?

**Comment by 3nids** [comment]
> Not the case for all files it seems, see [qgsprocessingparameter](https://github.com/qgis/QGIS/blob/6302af30a30eace54fc7c4673f6937b7527b2a1b/src/core/processing/qgsprocessingparameters.h#L338) for instance. Is it intentionnal ?

in this case, the whole TypeHeaderCode is manually generated. This can be switched indeed (I'd prefer leaving this for a follow-up).

> That mean that we would have a macro also for MappedType/ConvertFromTypeCode/End/ConvertToTypeCode see [qgsattributes](https://github.com/qgis/QGIS/blob/6302af30a30eace54fc7c4673f6937b7527b2a1b/src/core/qgsattributes.h#L146). ?

indeed, probably a macro

> I'm not sure about commenting them either as it is IMHO a little bit error prone. Comment followed with % would become SIP instruction and other comment would stay a comment.

indeed, that's why I plan to move these as macros. But for the sake of simplicity while we still have astyle running on some part of the code.

> can we not disable clang-format around given lines like with clang-tidy?

no, at least I failed at it. It's not really the formatting of these lines which are problematic but rather what happens after. The following method after a directive are not correctly formatted. In general clang-format seems a bit more picky than astyle and it's not super tolerant with our macros and injected sip code.

**Comment by troopa81** [comment]
> no, at least I failed at it. 

[This](https://releases.llvm.org/3.6.0/tools/clang/docs/ClangFormatStyleOptions.html#disabling-formatting-on-a-piece-of-code) doesn't work ?

**Comment by 3nids** [comment]
No. It works for the line in question, but messes up the following

**Comment by troopa81** [comment]
I'm not in favor of altering sip code either with comment or macros, it makes the whole thing even less readable IMHO. Most of sip code (all?) should not live in header file, it should be in different file included on the fly by sipify, but that's another story.

I won't block this PR but I'd rather wait for another contributor opinion.

**Comment by 3nids** [comment]
> I'm not in favor of altering sip code either with comment or macros, it makes the whole thing even less readable IMHO

We already had to do this for `%ConvertToTypeCode` (e.g. https://github.com/qgis/QGIS/blob/390ac1401314d12f25acba53330c92cc63b37e18/src/core/qgstemporalcontroller.h#L46)
So basically the same, but I get your point of altering the code.

> Most of sip code (all?) should not live in header file

I tend to find it useful to read the expected behavior in a single place, but that's debatable.

> I won't block this PR but I'd rather wait for another contributor opinion.

Thanks for the feedback, let's wait for more opinions :)

**Comment by nyalldawson** [comment]
I'm +1 to this -- I think the choices here look like good pragmatic solutions to a problem which does not have an ideal answer.

Regarding: 

>Most of sip code (all?) should not live in header file, it should be in different file included on the fly by sipify, but that's another story.

I'm personally in favour of this code living in the headers. It makes it really obvious when you're editing c++ classes what extra logic is involved in Python. If we siloed it away in a different place then it's very prone to bit rotting/getting out of sync.

**Comment by troopa81** [comment]
> We already had to do this for %ConvertToTypeCode

Fair point, so let's go for macros.

> I'm personally in favour of this code living in the headers. It makes it really obvious when you're editing c++ classes what extra logic is involved in Python. If we siloed it away in a different place then it's very prone to bit rotting/getting out of sync.

I think the opposite, QGIS should live without its binding the same way Qt live without its own (there is even 2 different ones) and so we should clearly separate both and maybe have some checker to avoid out-of-sync issues (not completely sure that's necessary).

Anyway, that's way out of the scope of this PR. LGTM (and replace comment with macro then :wink: )

**Review by troopa81 — APPROVED** [review]

**Pull request merged**